File: backend/app/api/v1/leads.py
"""Lead unsubscribe endpoint.

Public endpoint (no auth) for unsubscribing marketing leads.
Compatibility proxy to the Cloudflare D1-backed marketing Worker.
"""

from __future__ import annotations

import logging
import uuid

from fastapi import APIRouter, HTTPException, status

from app.core.config import settings
from app.services.marketing_worker import MarketingWorkerClient

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/leads", tags=["Leads"])


@router.get(
    "/unsubscribe",
    status_code=status.HTTP_200_OK,
)
async def unsubscribe_lead(lead_id: str) -> dict[str, bool]:
    """Forward a lead unsubscribe request to the marketing Worker."""
    try:
        uuid.UUID(lead_id)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="lead_id must be a valid UUID.",
        )

    marketing_worker = MarketingWorkerClient(
        base_url=settings.marketing_worker_url,
        secret=settings.marketing_worker_secret,
    )
    try:
        unsubscribed = await marketing_worker.unsubscribe(lead_id)
    except Exception as exc:
        logger.exception("Marketing Worker unsubscribe failed")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Marketing data service unavailable.",
        ) from exc

    if not unsubscribed:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lead not found.",
        )

    return {"success": True}

File: workers/api/src/routes/leads.ts
import { Hono } from 'hono'

import {
  MarketingWorkerNotConfiguredError,
  unsubscribeLead,
} from '../services/marketing-worker'
import type { AppBindings } from '../types'

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i

export interface LeadsRouteDependencies {
  fetch?: typeof fetch
}

export function createLeadsRoutes(
  dependencies: LeadsRouteDependencies = {},
): Hono<AppBindings> {
  const leads = new Hono<AppBindings>()

  leads.get('/unsubscribe', async (c) => {
    const leadId = c.req.query('lead_id') ?? ''
    if (!UUID_RE.test(leadId)) {
      return c.json({ detail: 'lead_id must be a valid UUID.' }, 422)
    }
    try {
      const unsubscribed = await unsubscribeLead(
        leadId,
        c.env,
        dependencies.fetch ?? fetch,
      )
      if (!unsubscribed) {
        return c.json({ detail: 'Lead not found.' }, 404)
      }
      return c.json({ success: true })
    } catch (error) {
      if (error instanceof MarketingWorkerNotConfiguredError) {
        return c.json({ detail: 'Marketing data service unavailable.' }, 503)
      }
      return c.json({ detail: 'Marketing data service unavailable.' }, 503)
    }
  })

  return leads
}

export const leadsRoutes = createLeadsRoutes()
